### ISSUE TITLE
Function start/2 has no local return under 1.5.0-dev (df7e0ca55), but no warning under 1.4.2

### ISSUE THREAD

**Issue by KronicDeth** [elixir-lang/elixir#6170]
### Environment

* Elixir & Erlang versions (elixir --version):

Elixir: 1.5.0-dev (df7e0ca55cd03e3d46f426c7cd02fd25dcf2df87) vs 1.4.2 (homebrew)
Erlang: 19.3 (homebrew)

* Operating system: macOS 10.12.5

### Current behavior

1. `git clone git@github.com:elixir-lang/elixir.git elixir-lang/elixir`
2. `pushd elixir-lang/elixir`
3. `git checkout df7e0ca55cd03e3d46f426c7cd02fd25dcf2df87`
4. `make clean`
5. `make test`
6. `export PATH=${PWD}/bin:${PATH}`
7. `elixir --version` is `1.5.0-dev (df7e0ca55)`
8. `popd`
9. `git clone git@github.com:C-S-D/calcinator.git`
10. `pushd C-S-D/calcinator`
11. `git checkout v3.0.0`
12. Make sure a previous version of elixir 1.5.0-dev PLT isn't being used: `rm -f ~/.mix/dialyze_erlang-19.3_elixir-1.5.0-dev.plt`
13. Make sure a previous version elixir 1.5.0-dev PLT and beam for project isn't being used: `rm -rf _build/test`
14. `mix deps.get`
15. `mix dialyze` (Wait a few minutes)

```
lib/calcinator/resources.ex:216: Guard test _@3::binary() == 'nil' can never succeed

test/support/calcinator/application.ex:8: Function start/2 has no local return

** (Mix) Dialyzer reported 2 warnings
```

The `lib/calcinator/resources.ex:216` is new since the last version of 1.5.0-dev I tried.

> ```elixir
>   for(potential_field <- fields(ecto_schema_module),
>         potential_field_string = to_string(potential_field),
>         potential_field_string == field_string, do: potential_field)
>     |> case do
>       [field] ->
>         {:ok, field}
>       [] ->
>         {:error, attribute}
>     end
> ```
> -- https://github.com/C-S-D/calcinator/blob/v3.0.0/lib/calcinator/resources.ex#L216-L224

The change that's persisted for the last few versions of `1.5.0-dev` is `test/support/calcinator/application.ex:8: Function start/2 has no local return`

> ```elixir
>   def start(_type, _args) do
>     import Supervisor.Spec, only: [supervisor: 2]
> 
>     children = [
>       supervisor(Calcinator.Resources.Ecto.Repo.Repo, [])
>     ]
>     opts = [strategy: :one_for_one, name: Calcinator.Supervisor]
>     Supervisor.start_link(children, opts)
>   end
> ```
> -- https://github.com/C-S-D/calcinator/blob/v3.0.0/test/support/calcinator/application.ex#L8-L16

### Expected behavior

1. Open a new tab (to clear environment `export`)
2. Use homebrew elixir 1.4.2: `elixir --version` is `1.4.2`
3. `rm -rf _build`
4. `mix dialyze` (wait a few minutes)

No warnings

### Notes

These could either be regressions in the types
* `Application.start/2`
* `Supervisor.start_link/2`

OR

It could be improvements to the types that are showing pre-existing bugs in how I have the `for` and `start/2` setup that just happened to work under 1.4.2.

I'm fine with either, but if it is pre-existing bugs in my code, I'm not sure from the errors how to figure out what the problem is.

**Comment by josevalim** [comment]
I have no idea why it is complaining that Supervisor.start_link has no return. However I do know why the first one is happening. Thanks for the report!

**Comment by josevalim** [comment]
Unfortunately I could not reproduce any of them in our suite. I will try Erlang 19.3.

**Comment by fishcakez** [comment]
The `start/2` no return is fixed in https://github.com/elixir-lang/elixir/commit/6e85a0f934f666771c4d5e48ac372673dae6be0b.

**Comment by fishcakez** [comment]
Minimal example test for `binary() == nil` in https://github.com/elixir-lang/elixir/tree/jf-for-bool-check. We could mark the case patterns as generated.

**Comment by josevalim** [comment]
The second one is fixed in 906fd46d512abee0ad7344b3057c6ae8ac7faf12. Thanks for the help @fishcakez! :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
